source/crc32.c

```diff
@@ -0,0 +1,99 @@
+#include <aws/cal/crc32.h>
+#include <aws/cal/hash.h>
+#include <aws/checksums/crc.h>
+#include <aws/common/byte_buf.h>
+#include <stdint.h>
+
+static int crc32_update(struct aws_hash *hash, const struct aws_byte_cursor *to_hash);
+static int crc32c_update(struct aws_hash *hash, const struct aws_byte_cursor *to_hash);
+static int s_finalize(struct aws_hash *hash, struct aws_byte_buf *output);
+static void s_destroy(struct aws_hash *hash) {
+    (void)hash;
+}
+/* can we reuse some of these as defaults defined in hash.c?*/
+
+/* is it possible to define this in a function somehow? I don't like that we define almost the same vtable twice */
+static struct aws_hash_vtable crc32_vtable = {
+    .destroy = s_destroy,
+    .update = crc32_update,
+    .finalize = s_finalize,
+    .alg_name = "CRC32",
+    .provider = "AWS", /* is this right? */
+};
+
+static struct aws_hash_vtable crc32c_vtable = {
+    .destroy = s_destroy,
+    .update = crc32c_update,
+    .finalize = s_finalize,
+    .alg_name = "CRC32C",
+    .provider = "AWS", /* is this right? */
+};
+
+static struct aws_hash *s_crc32_common_init(struct aws_allocator *allocator, struct aws_hash_vtable *vtable) {
+    struct aws_hash *hash = aws_mem_acquire(allocator, sizeof(struct aws_hash));
+
+    if (!hash) {
+        return NULL;
+    }
+
+    hash->allocator = allocator;
+    hash->vtable = vtable;
+    hash->impl = 0;
+    hash->digest_size = AWS_CRC32C_LEN;
+    hash->good = true;
+
+    return hash;
+}
+
+struct aws_hash *aws_crc32_default_new(struct aws_allocator *allocator) {
+    return s_crc32_common_init(allocator, &crc32_vtable);
+}
+
+struct aws_hash *aws_crc32c_default_new(struct aws_allocator *allocator) {
+    return s_crc32_common_init(allocator, &crc32c_vtable);
+}
+
+/* would it be useful to collect most of the logic into a helper function with a flag, and have two versions with the
```

Comment:
You could do this and take a function pointer as an argument, just don't do it as a branching (switch/if) implementation, that'll be slower.